crucible-llvm/src/Lang/Crucible/LLVM/SimpleLoopFixpoint.hs

```diff
@@ -0,0 +1,748 @@
+import           Data.Either
+import           Data.Foldable
+import qualified Data.IntMap as IntMap
+import           Data.IORef
+import qualified Data.List as List
+import           Data.Maybe
+import qualified Data.Map as Map
+import           Data.Map (Map)
+import qualified Data.Set as Set
+import qualified System.IO
+import           Numeric.Natural
+
+import qualified Data.BitVector.Sized as BV
+import           Data.Parameterized.Classes
+import qualified Data.Parameterized.Context as Ctx
+import qualified Data.Parameterized.Map as MapF
+import           Data.Parameterized.Map (MapF)
+import           Data.Parameterized.NatRepr
+import           Data.Parameterized.TraversableF
+import           Data.Parameterized.TraversableFC
+
+import qualified What4.Config as W4
+import qualified What4.Interface as W4
+
+import qualified Lang.Crucible.Analysis.Fixpoint.Components as C
+import qualified Lang.Crucible.Backend as C
+import qualified Lang.Crucible.CFG.Core as C
+import qualified Lang.Crucible.Panic as C
+import qualified Lang.Crucible.Simulator.CallFrame as C
+import qualified Lang.Crucible.Simulator.EvalStmt as C
+import qualified Lang.Crucible.Simulator.ExecutionTree as C
+import qualified Lang.Crucible.Simulator.GlobalState as C
+import qualified Lang.Crucible.Simulator.Operations as C
+import qualified Lang.Crucible.Simulator.RegMap as C
+import qualified Lang.Crucible.Simulator as C
+
+import qualified Lang.Crucible.LLVM.Bytes as C
+import qualified Lang.Crucible.LLVM.DataLayout as C
+import qualified Lang.Crucible.LLVM.MemModel as C
+import qualified Lang.Crucible.LLVM.MemModel.MemLog as C hiding (Mem)
+import qualified Lang.Crucible.LLVM.MemModel.Pointer as C
+import qualified Lang.Crucible.LLVM.MemModel.Type as C
+
+
+data FixpointStatus = BeforeFixpoint | ComputeFixpoint | CheckFixpoint | AfterFixpoint
+  deriving (Eq, Ord, Show)
+
+data FixpointEntry sym tp = FixpointEntry
+  { headerValue :: W4.SymExpr sym tp
+  , bodyValue :: W4.SymExpr sym tp
+  }
+
+instance OrdF (W4.SymExpr sym) => OrdF (FixpointEntry sym) where
+  compareF x y = case compareF (headerValue x) (headerValue y) of
+    LTF -> LTF
+    EQF -> compareF (bodyValue x) (bodyValue y)
+    GTF -> GTF
+
+instance OrdF (FixpointEntry sym) => W4.TestEquality (FixpointEntry sym) where
+  testEquality x y = case compareF x y of
+    EQF -> Just Refl
+    _ -> Nothing
+
+instance W4.TestEquality (FixpointEntry sym) => C.EqF (FixpointEntry sym) where
+  eqF x y = isJust $ W4.testEquality x y
+
+data MemFixpointEntry sym = forall w . (1 <= w) => MemFixpointEntry
+  { memFixpointEntrySym :: sym
+  , memFixpointEntryJoinVariable :: W4.SymBV sym w
+  }
+
+instance W4.TestEquality (W4.SymExpr sym) => Eq (MemFixpointEntry sym) where
+  (MemFixpointEntry _ x_join_variable) == (MemFixpointEntry _ y_join_variable)
+    | Just Refl <- W4.testEquality x_join_variable y_join_variable =
+      True
+    | otherwise = False
+
+data FixpointStuff sym blocks = forall args . FixpointStuff
```

Comment:
Can we use a more descriptive name than `Stuff`?

Reply:
Agreed, I'd like to know a lot more about what's going on this this type; it seems important.